### PR TITLE
feat: codecatalyst engine

### LINKED ISSUE #8: Engine: Implement initial version of CodeCatalyst Engine flavour
We would like to be able to generate a CodeCatalyst Workflow File if you select CodeCatalyst as an engine with the same features that we have for Github Actions today

### PR DESCRIPTION
**Pull request by Lock128**
Fixes #8 

This PR adds an initial version of the CodeCatalyst engine

### CODE REVIEW

File: .projenrc.ts [hunk 1]
```diff
@@ -53,8 +59,17 @@ const project = new cdk.JsiiProject({
       esModuleInterop: true,
     },
   },
+  gitignore: [
```
Comment (hoegertn) [inline .projenrc.ts:62]: Can you please elaborate why?

File: .projenrc.ts [hunk 2]
```diff
@@ -53,8 +59,17 @@ const project = new cdk.JsiiProject({
...
+    '!lib/',
+    '**/.DS_Store',
+  ],
 });
 
+project.package.addPackageResolutions('projen@0.86.7');
+
+project.tasks.tryFind('docgen')?.reset('echo no-execute on local');
```
Comment (hoegertn) [inline .projenrc.ts:70]: We should remove this before merging

File: .projenrc.ts [hunk 3]
```diff
@@ -53,8 +59,17 @@ const project = new cdk.JsiiProject({
       esModuleInterop: true,
     },
   },
+  gitignore: [
+    '!lib/',
+    '**/.DS_Store',
+  ],
 });
 
+project.package.addPackageResolutions('projen@0.86.7');
```
Comment (hoegertn) [inline .projenrc.ts:68]: This should go away; we are at projen 0.87.x

File: src/awscdk/codecatalyst.ts [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+- steps per stage - preInstall, preSynth, ... - DONE
+
+example: https://github.com/aws-community-dach/event-system-backend
+
+test docgen: https://github.com/open-constructs/aws-cdk-library
+
+
+*/
+
+export interface CodeCatalystIamRoleConfig {
```
Comment (hoegertn) [inline src/awscdk/codecatalyst.ts:38]: Cool, given that we need it for all engines and it behaves the same, I will refactor this once merged to the base class

File: src/awscdk/codecatalyst.ts [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+
+export class CodeCatalystCDKPipeline extends CDKPipeline {
+
+  public readonly needsVersionedArtifacts: boolean;
+
+  private deploymentWorkflowBuilder: WorkflowBuilder;
+  private environments: Map<String, Environment> = new Map();
+  private deploymentStages: string[] = [];
+
+  private readonly bp: Blueprint;
```
Comment (hoegertn) [inline src/awscdk/codecatalyst.ts:63]: Do we want to expose this so that users can add to the blueprint?

File: src/awscdk/codecatalyst.ts [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+
+    for (const stage of (options.independentStages ?? [])) {
+      this.createIndependentDeployment(stage);
+    }
+
+    const yml = new YamlFile(this, '.codecatalyst/workflows/deploy.yaml', {
+      obj: this.deploymentWorkflowBuilder.getDefinition(),
+
+    });
+    yml.synthesize();
```
Comment (hoegertn) [inline src/awscdk/codecatalyst.ts:98]: That should not be needed. projen calls all synth methods

File: src/awscdk/codecatalyst.ts [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+    for (const stage of options.stages) {
+      this.createDeployment(stage);
+    }
+
+    for (const stage of (options.independentStages ?? [])) {
+      this.createIndependentDeployment(stage);
+    }
+
+    const yml = new YamlFile(this, '.codecatalyst/workflows/deploy.yaml', {
+      obj: this.deploymentWorkflowBuilder.getDefinition(),
```
Comment (hoegertn) [inline src/awscdk/codecatalyst.ts:95]: Should this be lazy? So we can add stages etc after the constructor?

File: src/awscdk/codecatalyst.ts [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+    });
+  }
+
+  public createDeployment(stage: DeploymentStage): void {
+    let dependsOn = `deploy_${this.deploymentStages.at(-1)!}`;
+    if (stage.manualApproval === true) {
+      this.deploymentWorkflowBuilder.addGenericAction({
+        Identifier: 'aws/approval@v1',
+        actionName: `approve_${stage.name}`,
+        DependsOn: [`deploy_${this.deploymentStages.at(-1)!}`],
```
Comment (hoegertn) [inline src/awscdk/codecatalyst.ts:267]: use dependsOn variable

File: src/awscdk/codecatalyst.ts [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+    this.deploymentStages.push(stage.name);
+  }
+
+  public createIndependentDeployment(stage: DeploymentStage): void {
+    let dependsOn = 'PublishAssetsToAWS';
+    if (stage.manualApproval === true) {
+      this.deploymentWorkflowBuilder.addGenericAction({
+        Identifier: 'aws/approval@v1',
+        actionName: `approve_${stage.name}`,
+        DependsOn: [`deploy_${this.deploymentStages.at(-1)!}`],
```
Comment (hoegertn) [inline src/awscdk/codecatalyst.ts:309]: should be dependsOn with the publish step

File: src/steps/step.ts [hunk 10]
```diff
@@ -38,6 +38,24 @@ export interface GithubStepConfig {
...
+  readonly needs: string[];
+
+  /** Commands wrapped as GitHub Action job steps. */
+  readonly commands: string[];
+
+  /** Additional environment variables to set for this step. */
+  readonly env: { [key: string]: string };
+
+  /** Additional job permissions needed */
+  readonly permissions?: JobPermissions;
```
Comment (hoegertn) [inline src/steps/step.ts:56]: Should this be using the GitHub Model?

File: .projenrc.ts (hunk 1 above)
Comment (Lock128) [reply .projenrc.ts:62]: dont know why that sneaked in - removing

File: src/awscdk/codecatalyst.ts (hunk 5 above)
Comment (Lock128) [reply src/awscdk/codecatalyst.ts:63]: I don't think we should - as this is right now only used internally as thats the "workflow generator" - I dont think users should be "aware" of the underlying implementation

File: src/awscdk/codecatalyst.ts (hunk 6 above)
Comment (Lock128) [reply src/awscdk/codecatalyst.ts:98]: It doesnt for a YamlFile object - or at least I wasnt able to get it do that... As this is the "base" class implementation... maybe I would need to add this as a member of the class to make projen call it?

File: src/steps/step.ts (hunk 10 above)
Comment (Lock128) [reply src/steps/step.ts:56]: its not being used, c/p error - removed

File: .projenrc.ts (hunk 1 above)
Comment (hoegertn) [reply .projenrc.ts:62]: still there

**Review by hoegertn — APPROVED** [review]

**Pull request merged**